### ISSUE TITLE
Rendering issue: when panning, duplicate of graph sometimes stays static

### ISSUE THREAD

**Issue by b3nn0** [dbuezas/lovelace-plotly-graph-card#124]
Hi,
See gif below. It seems to have started recently, around the time when `period: auto` was introduced, but can't say a specific version.

Card definition:
```
type: custom:plotly-graph
layout:
  height: 400
entities:
  - entity: sensor.wolf_aussentemperatur
    line:
      shape: linear
  - entity: sensor.dachboden_temperatur
    line:
      shape: linear
  - entity: sensor.bme680_temperature
    name: Bürotemperatur
    line:
      shape: linear
  - entity: sensor.wohnzimmertemperatur
    line:
      shape: linear
defaults:
  entity:
    show_value:
      right_margin: 15
  yaxes:
    fixedrange: true
refresh_interval: 60
hours_to_show: 48
title: Außentemperatur

```

Issue:
![graph](https://user-images.githubusercontent.com/1858945/200262776-a56bf3a2-f787-46cb-a344-307e79239656.gif)

Some hints:
- It happens in both, Firefox and Chromium
- It doesn't happen for all my graphs, only some, and always for the same (i.e. always for the red one in above example, but never the others)
- It also happens when zooming

EDIT: Just noticed: `period: auto` seems to have stopped working for me in the latest versions, too. Only shows daily all the time now. If I explicitly set the period definitions for different zooms like in the readme, it works fine. Should I open another issue for that?

EDIT2: Just retested with older versions. In 1.5.3 everything is fine, in 1.6.0 the issue appears.

**Comment by dbuezas** [comment]
Hi @b3nn0, thanks for letting me know.
Does this also occur in the latest version (1.6.3)?
I can't reproduce this bug so I'll need the data that breaks it, you can do the following to share it with me:

```yaml
entities:
  - entity: sensor.the-problematic-one
    ...whatever else you have configured...
    lambda: |-
       (ys, xs) => {
          const data = {x: xs, y: ys};
          console.log("sensor.the-problematic-one", data);
          return data
       }
```
Then when you see the issue in the screen, [open the dev tools](https://balsamiq.com/support/faqs/browserconsole/), go to the console, right click on the object containing the x and y data and `copy object`.
Then you can paste it here.
With that I'll be able to produce the bug in my computer and look for the root cause.

**Comment by b3nn0** [comment]
Here is the data:
```
{
  "x": [
    "2022-11-05T10:41:24.414Z",
    "2022-11-05T11:31:19.438Z",
    "2022-11-05T11:46:19.447Z",
    "2022-11-05T11:58:19.419Z",
    "2022-11-05T11:59:19.420Z",
    "2022-11-05T12:16:19.358Z",
    "2022-11-05T14:04:19.199Z",
    "2022-11-05T14:05:19.188Z",
    "2022-11-05T14:06:19.196Z",
    "2022-11-05T14:12:19.221Z",
    "2022-11-05T14:13:19.228Z",
    "2022-11-05T14:14:19.174Z",
    "2022-11-05T14:15:19.179Z",
    "2022-11-05T14:42:19.143Z",
    "2022-11-05T14:52:19.138Z",
    "2022-11-05T15:41:19.058Z",
    "2022-11-05T17:12:18.978Z",
    "2022-11-05T17:13:18.983Z",
    "2022-11-05T17:14:18.930Z",
    "2022-11-05T17:18:18.924Z",
    "2022-11-05T17:19:18.920Z",
    "2022-11-05T17:21:18.934Z",
    "2022-11-05T17:22:18.943Z",
    "2022-11-05T18:03:18.863Z",
    "2022-11-05T18:19:18.851Z",
    "2022-11-05T18:37:18.871Z",
    "2022-11-05T20:00:18.745Z",
    "2022-11-05T20:02:18.688Z",
    "2022-11-05T20:30:18.659Z",
    "2022-11-05T20:46:18.670Z",
    "2022-11-05T20:48:18.676Z",
    "2022-11-05T20:50:18.627Z",
    "2022-11-05T22:28:18.492Z",
    "2022-11-05T22:29:18.486Z",
    "2022-11-05T22:30:18.492Z",
    "2022-11-06T00:20:18.355Z",
    "2022-11-06T00:21:18.373Z",
    "2022-11-06T00:22:18.373Z",
    "2022-11-06T00:26:18.321Z",
    "2022-11-06T00:27:18.327Z",
    "2022-11-06T00:29:18.322Z",
    "2022-11-06T00:56:18.312Z",
    "2022-11-06T01:21:18.282Z",
    "2022-11-06T01:22:18.297Z",
    "2022-11-06T05:01:17.935Z",
    "2022-11-06T05:04:17.937Z",
    "2022-11-06T05:38:17.886Z",
    "2022-11-06T05:39:17.878Z",
    "2022-11-06T05:40:17.888Z",
    "2022-11-06T05:41:17.891Z",
    "2022-11-06T05:50:17.864Z",
    "2022-11-06T05:52:17.867Z",
    "2022-11-06T06:01:17.847Z",
    "2022-11-06T06:02:17.851Z",
    "2022-11-06T08:49:17.610Z",
    "2022-11-06T08:50:17.608Z",
    "2022-11-06T08:54:17.747Z",
    "2022-11-06T08:55:37.024Z",
    "2022-11-06T08:55:37.959Z",
    "2022-11-06T08:55:37.959Z",
    "2022-11-06T08:59:42.621Z",
    "2022-11-06T08:59:43.207Z",
    "2022-11-06T09:00:17.590Z",
    "2022-11-06T09:02:47.384Z",
    "2022-11-06T09:02:48.041Z",
    "2022-11-06T09:02:48.041Z",
    "2022-11-06T09:12:23.942Z",
    "2022-11-06T09:12:24.658Z",
    "2022-11-06T09:45:17.590Z",
    "2022-11-06T10:04:17.521Z",
    "2022-11-06T10:05:17.529Z",
    "2022-11-06T12:03:17.359Z",
    "2022-11-06T12:22:17.318Z",
    "2022-11-06T12:42:17.327Z",
    "2022-11-06T13:13:17.243Z",
    "2022-11-06T13:17:17.275Z",
    "2022-11-06T13:18:17.278Z",
    "2022-11-06T13:40:17.229Z",
    "2022-11-06T13:43:17.184Z",
    "2022-11-06T13:44:17.193Z",
    "2022-11-06T13:45:17.194Z",
    "2022-11-06T13:47:17.191Z",
    "2022-11-06T13:50:17.205Z",
    "2022-11-06T13:58:17.174Z",
    "2022-11-06T15:40:17.072Z",
    "2022-11-06T15:50:17.036Z",
    "2022-11-06T15:59:17.007Z",
    "2022-11-06T16:33:16.959Z",
    "2022-11-06T16:37:16.968Z",
    "2022-11-06T17:26:16.921Z",
    "2022-11-06T22:25:16.501Z",
    "2022-11-06T23:16:16.389Z",
    "2022-11-06T23:18:16.393Z",
    "2022-11-06T23:35:16.387Z",
    "2022-11-06T23:36:16.390Z",
    "2022-11-06T23:38:16.352Z",
    "2022-11-06T23:40:16.353Z",
    "2022-11-06T23:41:16.360Z",
    "2022-11-07T01:31:16.201Z",
    "2022-11-07T01:41:16.188Z",
    "2022-11-07T01:43:16.190Z",
    "2022-11-07T02:10:36.981Z",
    "2022-11-07T02:11:42.366Z",
    "2022-11-07T03:49:16.034Z",
    "2022-11-07T03:51:16.028Z",
    "2022-11-07T03:52:16.042Z",
    "2022-11-07T03:53:16.046Z",
    "2022-11-07T03:55:16.003Z",
    "2022-11-07T04:03:16.021Z",
    "2022-11-07T04:04:16.021Z",
    "2022-11-07T04:08:15.974Z",
    "2022-11-07T04:09:15.982Z",
    "2022-11-07T05:22:15.876Z",
    "2022-11-07T05:45:15.848Z",
    "2022-11-07T05:47:15.842Z",
    "2022-11-07T07:01:15.749Z",
    "2022-11-07T07:39:15.714Z",
    "2022-11-07T07:53:15.721Z",
    "2022-11-07T08:00:15.668Z",
    "2022-11-07T08:56:15.571Z",
    "2022-11-07T08:57:15.573Z",
    "2022-11-07T09:08:53.301Z",
    "2022-11-07T09:08:54.076Z",
    "2022-11-07T10:09:15.471Z",
    "2022-11-07T10:41:57.605Z"
  ],
  "y": [
    "21.3",
    "21.4",
    "21.5",
    "21.4",
    "21.5",
    "21.4",
    "21.3",
    "21.4",
    "21.3",
    "21.4",
    "21.3",
    "21.4",
    "21.3",
    "21.4",
    "21.3",
    "21.2",
    "21.1",
    "21.2",
    "21.1",
    "21.2",
    "21.1",
    "21.2",
    "21.1",
    "21.2",
    "21.3",
    "21.4",
    "21.5",
    "21.4",
    "21.5",
    "21.4",
    "21.5",
    "21.4",
    "21.3",
    "21.4",
    "21.3",
    "21.2",
    "21.3",
    "21.2",
    "21.3",
    "21.2",
    "21.3",
    "21.2",
    "21.3",
    "21.2",
    "21.1",
    "21.2",
    "21.1",
    "21.2",
    "21.1",
    "21.2",
    "21.1",
    "21.2",
    "21.1",
    "21.2",
    "21.1",
    "21.0",
    "21.1",
    null,
    "21.1",
    "21.1",
    null,
    "21.1",
    "21.2",
    null,
    "21.2",
    "21.2",
    null,
    "21.2",
    "21.3",
    "21.2",
    "21.3",
    "21.4",
    "21.5",
    "21.6",
    "21.5",
    "21.6",
    "21.5",
    "21.6",
    "21.8",
    "21.7",
    "21.8",
    "21.7",
    "21.6",
    "21.5",
    "21.6",
    "21.5",
    "21.6",
    "21.7",
    "21.6",
    "21.5",
    "21.4",
    "21.3",
    "21.4",
    "21.3",
    "21.4",
    "21.3",
    "21.4",
    "21.3",
    "21.2",
    "21.3",
    "21.2",
    null,
    "21.2",
    "21.1",
    "21.2",
    "21.1",
    "21.2",
    "21.1",
    "21.2",
    "21.1",
    "21.2",
    "21.1",
    "21.2",
    "21.1",
    "21.2",
    "21.3",
    "21.4",
    "21.3",
    "21.4",
    "21.3",
    "21.4",
    null,
    "21.4",
    "21.5",
    "21.5"
  ]
}
```

For this, I scrolled the graph just like in the gif and then copied the newest data after verifying that it showed the same symptoms.

**Comment by dbuezas** [comment]
I managed to reproduce it. The issue is the same as the last one but in the extremes: if the plot starts with one valid data point and is followed by unavailable, plotlyjs gets confused. I'll release the fix soon

**Comment by dbuezas** [comment]
I forgot: thanks for the data

**Comment by b3nn0** [comment]
Nah, thank YOU for doing all this - this has become by far the best graph card for HA.

**Comment by b3nn0** [comment]
Unfortunately, in 1.7.2 the issue still seems to occur.. :-(
Some more observations:
On initially loading the dashboard, the lambda is executed 4 times.
Data this time: https://pastebin.com/tj3JU1YZ

strangely enough, it always happens for this graph, but never for the others. I also tried to randomly change the config (refresh_interval, add/remove some other entities, reorder, ...) but that doesn't change anything.

**Comment by dbuezas** [comment]
How annoying... can you remove the "show_value" and tell me if that makes a difference?
Also another gif, it gives me some insights on what happens on the edges of the "ghost"

**Comment by b3nn0** [comment]
![graph](https://user-images.githubusercontent.com/1858945/200542614-a96e8c06-aaa3-462b-8a96-d2ad81425e1f.gif)

This time even managed to tripple (quadruple?) the graph..

It's not too critical though IMO.. not beautiful, but I guess there are more critical things in the world..

**Comment by b3nn0** [comment]
Ah, just tried it with a thick, semi transparent line, and indeed, the graph already breaks in the long-drag-to-the-left I did in 2. above:
![graph](https://user-images.githubusercontent.com/1858945/200619039-ab99b1d6-f8b2-4185-b622-61b43920785d.gif)

**Comment by dbuezas** [comment]
Uhm. I can't reproduce it this time. 
Let's try a couple of things:
1. open the browser's [devtools](https://balsamiq.com/support/faqs/browserconsole) and check that the version is correct:
![image](https://user-images.githubusercontent.com/777196/200627016-a8a422c1-b720-4c42-8df0-6d23eeb8c69e.png)
2. Try with this lambda 
```yaml
lambda: |-
  ys => ys.filter(y => y || 0 ) # will replace the gaps with zero
```
This will remove all gaps. If the issue goes away, we'd have confirmed the issue is still the gaps.
3. Try this other lambda:
```yaml
lambda: |-
  (ys,xs) => {
    for (let i = 0; i < xs.length; i++) {
      if (!isDefined(ys[i - 1]) && isDefined(ys[i]) && !isDefined(ys[i + 1])) {
        ys.splice(i, 0, ys[i]);
        xs.splice(i, 0, xs[i] + 1);
      }
  }
  return {x:xs, y:ys};
}
```
This is like the existing fix, except that it avoids repeating the exact same timestamp when adding fake duplicate values (`xs[i] + 1`)

**Comment by b3nn0** [comment]
1. Yes, it does say 1.7.2 production.
2. This fixes the issue, but there are still gaps for some reason. Additionally, it causes some other strange effects. I.e. while scrolling, some gaps are small. When scrolling further, they suddenly change in size?! Strange things happening..
![graph](https://user-images.githubusercontent.com/1858945/200636333-43517622-6782-43ec-b8c6-1ad7a4c6ba97.gif)

I changed the lambda a bit to print the ys, and indeed some of them read the string value "unknown". No idea where that's coming from, as it wasn't seen in the original value-dumps for some reason...?

Filtering "unknown" values as well will then indeed remove the gaps _and_ the duplicated lines.

3. This lambda does not fix the issue - looks just like before.


Further investigation of the 'unknown' values:
I checked my HASS DB:
`select * from states where entity_id='sensor.wolf_aussentemperatur' order by state_id desc limit 10000;`
and indeed, there are some 'unknown' values in there. No idea where they are coming from.
The data is received from some niche gas-heater reader via HASS's MQTT integration - maybe the source publishes them sometimes.... 

I finally tried one more lambda:
```
(ys) => ys.filter(y => y != "unknown" ? y : null)
```

This not only makes everything work correctly, it also removes all the gaps from the original graph. Not sure why.. I assume because there were never two consecutive null/unknown-values in the data, so plotly wouldn't know how big the gap should be?

tl;dr: Bad data, not your fault. Sorry!
Maybe it would make sense to filter out all non-numeric values initially to make sure others don't fall into the same trap...

**Comment by dbuezas** [comment]
i made a mistake in the lambda, I meant to write this:
```Yaml
lambda: |-
  ys => ys.map(y => y || 0 ) # map, not filter
```

**Comment by b3nn0** [comment]
Aah, that makes more sense. That one does fix it (and is equivalent tobmy last one I guess)

**Comment by dbuezas** [comment]
yes, true, they are mostly equivalent

**Comment by dbuezas** [comment]
> Yes, it does say 1.7.2 production.

ok, thanks for checking

> This fixes the issue, but there are still gaps for some reason. Additionally, it causes some other strange effects. I.e. while scrolling, some gaps are small. When scrolling further, they suddenly change in size?! Strange things happening..
graph

This is because `filter` in the `ys` array changes which `y` corresponds to which `x`. I meant to write `map`, not `filter`. Ups! 

> I changed the lambda a bit to print the ys, and indeed some of them read the string value "unknown".

I think I remember seeing that too at some point long ago. Maybe I should map that one to null too (so it creates a gap)

> 3. This lambda does not fix the issue - looks just like before.

This is because I wasn't paying much attention, the code doesn't run, as it calls a function that isn't there. Sorry for that

With the rest of your experiments, I think you confirmed that the issue are the gaps, but coming from other non-null values. What you made in your code makes a lot of sense, I think I need to extend my checks and convert both "unavailable" and "unknown" to null, then the "lonely datapoints" fix will take care of making the ghosts disappear. 

You mentioned that this `if (ys[i] == 'unknown') ys[i] = null` brings the bug back, that's odd. Maybe there is yet another form of gap. Please try this lambda:
```js
(ys) => ys.map(y => Number.isNaN(+y)?null:y)
```
and maybe output to the console all values that produce a NaN, to check there is not yet another odd value:
```js
(ys) => {
        for (let i = 0; i < ys.length; i++)
          if (Number.isNaN(+ys[i])) console.log(ys[i])
        return ys
      }
```


Ghosts, lonely datapoints... this is getting esoteric.

**Comment by b3nn0** [comment]
Awesome, this indeed fixes it completely: ` (ys) => ys.map(y => Number.isNaN(+y)?null:y) `

With the other one, the only thing I occasionally see is 'unknown', so I guess that was the culprit.

**Comment by dbuezas** [comment]
Great!
Although if "unknown" is the only one left, this experiment of yours should have actually had the same effect:
```yaml
for (let i = 0; i < ys.length; i++)
          if (ys[i] == 'unknown') ys[i] = null
```
I'll make a fix release soon anyway :)

**Comment by dbuezas** [comment]
note to self: raise an issue in plotly once this is fixed

**Comment by dbuezas** [comment]
You can now remove the "unknown" filter lambda 🎉  (update to 1.7.3)

**Comment by dbuezas** [comment]
Plotly issue: https://github.com/plotly/plotly.js/issues/6407
